File: apps/game_core/lib/game_core/worldgen.ex
defmodule GameCore.Worldgen do
  @moduledoc """
  Compile-time, deterministic placement of Resource nodes per Chunk.
  Pure function `resource_nodes/1` — given a chunk coord, returns the
  list of `{type, x, y}` entries that define the chunk's content. The
  DB only stores depletion state, never positions.

  v1 places a handful of Trees in a fixed pattern around each chunk's
  center. Procedural density / multi-type generation arrive in a later
  phase.
  """

  alias GameCore.ChunkGeometry

  @type node_spec :: %{type: :tree, x: integer(), y: integer()}

  # Tree offsets from chunk-center, in tenths of a world unit (× 100
  # sub-units). Tight cluster so a Player spawning at chunk-center has
  # at least one tree inside @interact_range_sq = 1.0u².
  @tree_offsets [
    {500, 500},
    {500, -500},
    {-500, 500},
    {-500, -500},
    {0, 0}
  ]

  @spec resource_nodes(ChunkGeometry.coord()) :: [node_spec()]
  def resource_nodes({cx, cy}) do
    size = ChunkGeometry.chunk_size()
    half = div(size, 2)
    center_x = cx * size + half
    center_y = cy * size + half

    for {dx, dy} <- @tree_offsets do
      %{type: :tree, x: center_x + dx, y: center_y + dy}
    end
  end

  @type portal_spec :: %{
          type: :dungeon,
          direction: :into_instance,
          x: integer(),
          y: integer()
        }

  @doc """
  Deterministic placement of Overworld Portals for `coord`. v1 places one
  `:dungeon` Portal in chunk `{0, 0}` and `[]` elsewhere. The Portal sits
  at a fixed offset from chunk-center so Players, who spawn at
  chunk-center, don't immediately overlap it on first connect.
  """
  @spec portals(ChunkGeometry.coord()) :: [portal_spec()]
  def portals({0, 0}) do
    size = ChunkGeometry.chunk_size()
    quarter = div(size, 4)
    [%{type: :dungeon, direction: :into_instance, x: quarter, y: quarter}]
  end

  def portals(_), do: []
end

File: sim/src/worldgen.rs
//! Deterministic, compile-time placement of Resource nodes and Portals per
//! chunk. Mirrors the Elixir `GameCore.Worldgen` and `GameCore.InstanceWorldgen`
//! exactly — same offsets, same coordinates — so the world is byte-identical.

use crate::components::{PortalDirection, PortalKind, ResourceKind};
use crate::geometry::{chunk_center, ChunkCoord, CHUNK_SIZE};

/// A worldgen-placed resource node.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct NodeSpec {
    pub kind: ResourceKind,
    pub x: i64,
    pub y: i64,
}

/// A worldgen-placed portal.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PortalSpec {
    pub kind: PortalKind,
    pub direction: PortalDirection,
    pub x: i64,
    pub y: i64,
}

/// Tree offsets from chunk-center, in sub-units. Tight cluster so a player
/// spawning at chunk-center has a tree inside interact range.
const TREE_OFFSETS: [(i64, i64); 5] = [
    (500, 500),
    (500, -500),
    (-500, 500),
    (-500, -500),
    (0, 0),
];

/// Overworld Resource nodes for `coord`: five trees around chunk-center.
pub fn resource_nodes(coord: ChunkCoord) -> Vec<NodeSpec> {
    let (cx, cy) = chunk_center(coord);
    TREE_OFFSETS
        .iter()
        .map(|&(dx, dy)| NodeSpec { kind: ResourceKind::Tree, x: cx + dx, y: cy + dy })
        .collect()
}

/// Overworld Portals for `coord`: one `:into_instance` dungeon portal in chunk
/// `{0,0}` at a quarter-offset from the origin, `[]` elsewhere.
pub fn portals(coord: ChunkCoord) -> Vec<PortalSpec> {
    if coord == ChunkCoord::new(0, 0) {
        let quarter = CHUNK_SIZE / 4;
        vec![PortalSpec {
            kind: PortalKind::Dungeon,
            direction: PortalDirection::IntoInstance,
            x: quarter,
            y: quarter,
        }]
    } else {
        Vec::new()
    }
}

/// Instance-local Portals for `coord`: one `:out_of_instance` return portal at
/// the center of chunk `{1,1}` (middle of the 3×3 grid), `[]` elsewhere.
pub fn instance_portals(coord: ChunkCoord) -> Vec<PortalSpec> {
    if coord == ChunkCoord::new(1, 1) {
        let (x, y) = chunk_center(coord);
        vec![PortalSpec {
            kind: PortalKind::Dungeon,
            direction: PortalDirection::OutOfInstance,
            x,
            y,
        }]
    } else {
        Vec::new()
    }
}

/// Center of Instance chunk `{1,1}` in instance-local sub-units — where the
/// return portal sits.
pub fn return_portal_pos() -> (i64, i64) {
    chunk_center(ChunkCoord::new(1, 1))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn five_trees_around_center() {
        let nodes = resource_nodes(ChunkCoord::new(0, 0));
        assert_eq!(nodes.len(), 5);
        // Center tree is exactly at chunk center.
        assert!(nodes.iter().any(|n| n.x == 8_000 && n.y == 8_000));
        assert!(nodes.iter().any(|n| n.x == 8_500 && n.y == 8_500));
    }

    #[test]
    fn portal_only_in_origin_chunk() {
        let p = portals(ChunkCoord::new(0, 0));
        assert_eq!(p.len(), 1);
        assert_eq!((p[0].x, p[0].y), (4_000, 4_000));
        assert_eq!(p[0].direction, PortalDirection::IntoInstance);
        assert!(portals(ChunkCoord::new(1, 0)).is_empty());
    }

    #[test]
    fn instance_return_portal_at_center() {
        let p = instance_portals(ChunkCoord::new(1, 1));
        assert_eq!(p.len(), 1);
        assert_eq!((p[0].x, p[0].y), (24_000, 24_000));
        assert_eq!(return_portal_pos(), (24_000, 24_000));
        assert!(instance_portals(ChunkCoord::new(0, 0)).is_empty());
    }
}
